rfcs/20190904-tfx-generic-container-based-component.md

```diff
@@ -0,0 +1,863 @@
+        '--output_examples', ' {{ outputs.examples.uri }}',
+        '--num_csv_columns', ' {{ exec_props.num_csv_columns }}' ],
+  platform_config=...
+)
+
+
+# Container-based ExampleGen's component spec.
+# Notice that this is similar to FileBasedExampleGenSpec,
+# but with a different set of PARAMETERS.
+class MyContainerBasedExampleGenSpec(ComponentSpec):
+  """ComponentSpec to drive my_example_gen.py as a Component."""
+  # Input artifacts.
+  INPUTS = {
+      "raw_input": ChannelParameter(type=standard_artifacts.ExternalArtifact),
+  }
+
+  # Output artifacts.
+  OUTPUTS = {
+      "examples": ChannelParameter(type=standard_artifacts.Examples),
+  }
+
+  # Parameters.
+  PARAMETERS = {
+      "num_csv_columns": ExecutionParameter(type=int),
+  }
+```
+
+*   `ExecutorSpec` and `ComponentSpec` for `my_trainer.py`
+
+```python
+# Container-based trainer's executor spec
+my_trainer_exec_spec = ExecutorContainerSpec(
+  container_image='gcr.io/my_project/my_trainer:stable',
+  command=['python']
+  args=['my_trainer.py',
+        '--input_examples', '{{ inputs.my_inputs.uri }}',
+        '--output_model', '{{ outputs.my_model.uri }}',
+        '--train_steps', '{{ exec_props.train_steps }}',]
+  # Platform config would specify use of GPU node pool for k8s, for example.
+  platform_config = ...
+)
+
+# Container-based trainer's component spec.
+# Notice that this is quite different from TrainerSpec, because of
+# the command line flags that my_trainer.py takes are different from what
+# TFX stock trainer takes. Nevertheless, it does produce an instance of
+# Model artifacts, which can then be consumed by downstream components.
+class MyContainerBasedTrainerSpec(ContainerBasedComponentSpec):
+  """ComponentSpec to drive my_trainer.py as a component."""
+  # Input artifacts.
+  INPUTS = {
+      "my_input": ChannelParameter(type_name=standard_artifacts.Examples),
+  }
+
+  # Output artifacts.
+  OUTPUTS = {
+      "my_model": ChannelParameter(type_name=standard.Artifacts.Model),
+  }
+
+  # Parameters
+  PARAMETERS = {
+      # Execution properties.
+      "train_steps": ExecutionParameter(type=int),
+  }
+```
+
+### Component definition based on a generic containerized program
+
+With the introduction of `ExecutorContainerSpec`, the way to define a component
+based on a containerized application is no different from any other custom
+component. Below are illustrations to define the components from the previous
+section in full, and their use in an end-to-end pipeline.
+
+*   Component definitions
+
+```python
+
+class MyContainerBasedExampleGen(BaseComponent):
+  """Wraps my_example_gen.py."""
+  SPEC_CLASS = MyContainerBasedExampleGenSpec
+
+  EXECUTOR_SPEC = my_example_gen_exec_sepc
+
+  # Optionally, if custom driver behavior is desired, such as checking
+  # mtime for file updates, one can define a custom Driver class to control
+  # the behavior of my_exmaple_gen.py inside the container.
+  DRIVER_CLASS = ...
```

Comment:
from naming perspective, would it be better to use the following for consistency?
COMP_SPEC (specification for the component) 
EXECUTOR_SPEC (specification for the executor)
DRIVER_SPEC (specification of the driver)

Current SPEC_CLASS and EXECUTOR_SPEC gives different meanings to "SPEC", might cause confusion.

Reply:
@ruoyu90 and @charlesccychen to comment on this. 

I believe there was a legitimate reason why ComponentSpec is a class, but EXECUTOR_SPEC and DRIVER_SPEC are instances.

Reply:
The ComponentSpec defines the input / output properties of a component, and an instance of that class can also be created to hold concrete values of those properties.  So the ComponentSpec class defines the execution schema while instances hold concrete instantiations.  Given this, it makes sense to have concrete ComponentSpec classes for each type of component.